modules/issue_tracker/php/endpoints/attachment.class.inc

```diff
@@ -0,0 +1,245 @@
+        }
+    }
+
+    /**
+     * Handle GET requests for attachment
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request.
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    private function _handleGET(ServerRequestInterface $request) : ResponseInterface
+    {
+        // Parse GET query params.
+        $values = $request->getQueryParams();
+
+        $file_hash = $values['file_hash'];
+        $filename  = $values['filename'];
+        $mime_type = $values['mime_type'];
+
+        $config = \NDB_Config::singleton();
+        $attachment_data_dir = rtrim(
+            $config->getSetting('IssueTrackerDataPath'),
+            '/'
+        );
+
+        $fileToDownload = trim($attachment_data_dir) .
+            '/attachments/' .
+            $file_hash;
+
+        if (!is_readable($fileToDownload)) {
+            throw new \LorisException('Forbidden Access');
+        }
+
+        $name = rawurldecode($filename);
+
+        return (new \LORIS\Http\Response())
+            ->withHeader('Content-Type', $mime_type)
+            ->withHeader(
+                'Content-Disposition',
+                'attachment; filename=' . basename($name)
+            )
+            ->withStatus(200)
+            ->withBody(new \LORIS\Http\FileStream($fileToDownload));
+    }
+
+    /**
+     * Handle POST requests for attachment
+     * and used for uploading attachment file.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request.
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    private function _handlePOST(ServerRequestInterface $request) : ResponseInterface
+    {
+        // Parse POST request body.
+        $values     = $request->getQueryParams();
+        $user       = \NDB_Factory::singleton()->user();
+        $attachment = new \LORIS\issue_tracker\UploadHelper();
+        $response   = $attachment->setupUploading(
+            $user,
+            $_FILES,
+            $values
+        );
+        return new \LORIS\Http\Response\JsonResponse(
+            $response
+        );
+    }
+
+    /**
+     * Handle DELETE requests for attachment
+     * and marks attachment as deleted from issue_tracker.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request.
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    private function _handleDELETE(ServerRequestInterface $request)
+    : ResponseInterface
+    {
+        // Verify User can delete file.
+        if ($this->_deletePermission($request)) {
+            // Parse request body.
+            $values    = $request->getQueryParams();
+            $file_hash = $values['file_hash'];
+            $factory   = \NDB_Factory::singleton();
+            $DB        = $factory->database();
+            $DB->update(
+                'issues_attachments',
+                array('deleted' => 1),
+                array('file_hash' => $file_hash)
```

Comment:
@xlecours the file_hash is used specifically for the feature of being able to upload attachments (files) to the issue. Multiple users can upload attachments to a single issue. The files should be stored only once on the server and multiple issues can have the same files be added as attachments. I wrote the code this way because it's the best way to go about it after thinking about it for weeks and I like it.

Reply:
resolved